rfd/0163-vnet.md

```diff
@@ -0,0 +1,416 @@
+
+To do this they can set the `spec.vnet_addr` field in the app resource to
+`api.example.private`, but VNet will also need to know to configure itself as
+the DNS server for `*.example.private`, and what to do with other DNS requests
+matching `*.example.private` with no matching Teleport app.
+
+To accomodate this the Teleport admin can configure a custom DNS zone
+cluster-wide with a new `vnet` Teleport resource:
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  custom_dns_zones:
+  - suffix: .example.private
+    upstream_nameservers:
+    - 1.2.3.4
+    - 4.5.6.7
+  - suffix: .test.private
+```
+
+In this example, whenever the user is logged in to the cluster where this `vnet`
+configuration resource is present, VNet will install itself as the DNS
+nameserver for `*.example.private` and `*.test.private`.
+
+### DNS queries and IP address assignment
+
+This design tries to avoid making the VNet client maintain an up-to-date list of
+all apps in all clusters with assigned IPs.
+To that end, IPs will be dynamically assigned to apps the first time DNS is
+queried for that app.
+
+The VNet process will regulary poll for the currently logged-in root and leaf
+clusters to discover their proxy addresses and
+[custom DNS zones](#custom-dns-zones).
+When any DNS query comes the process is as follows:
+
+1. If an IP address is already assigned to the full queried name, an
+   authoritative answer will be sent with that IP.
+1. If the name matches any `<app-name>.<proxy-address>.internal` we will query
+   the cluster to see if the app actually exists.
+    1. If the app does not exist, the DNS server will respond with an
+       authoritative non-answer to indicate this is not a valid name.
+    1. If the app does exist, a free IP will be assigned to it and it will be
+       returned in an authoritative answer.
+1. If the name matches `*.<custom-dns-zone>`:
+    1. VNet will determine which cluster(s) the matching custom DNS zone is in from
+       its local state.
+    1. VNet will query the cluster(s) apps to see if any have a `spec.vnet_addr`
+       matching the full name. This will use a predicate expression to do the
+       filtering server-side and avoid listing all apps.
+    1. If a match is found a free IP will be assigned to that app and it will be
+       returned in an authoritative answer.
+    1. If no match is found and the zone has no `upstream_nameservers` configured,
+       no answer will be sent.
+    1. The full DNS query will be forwarded to one of the `upstream_nameservers`,
+       selected randomly from the list, and any upstream response will be
+       returned to the client verbatim.
+       We may implement caching of upstream responses, in some cases the
+       downstream resolver will do its own caching.
+1. If the name did not match any `<proxy-address>.internal` or custom DNS zone,
+   no response will be sent.
+
+TODO: sequence diagram
+
+### IP range
```

Comment:
I'd be interested in seeing a short explanation as to why we would prefer to go with IPv4, rather than say, IPv6? It'd be fairly trivial for Teleport to obtain some IPv6 address space and reserve it for this purpose and avoid the risk of conflicting routing to the CGNAT range.

Reply:
Yeah the plan is to support both, I just didn't figure out how to assign IPv6 IPs yet, will be adding it. I'm not sure we can avoid IPv4 though in case some clients don't support IPv6